File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/utils/RequeryMetamodel.java
package com.coupang.springframework.data.requery.utils;

import io.requery.meta.Attribute;
import io.requery.meta.EntityModel;
import io.requery.meta.Type;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.HashSet;
import java.util.Optional;
import java.util.Set;

/**
 * RequeryMetamodel
 *
 * @author debop@coupang.com
 * @since 18. 6. 7
 */
public class RequeryMetamodel {

    private final EntityModel entityModel;
    private Optional<Collection<Class<?>>> managedTypes = Optional.empty();


    public RequeryMetamodel(@NotNull EntityModel entityModel) {
        this.entityModel = entityModel;
    }

    public boolean isRequeryManaged(@NotNull Class<?> entityClass) {
        return getManagedTypes().contains(entityClass);
    }

    public boolean isSingleIdAttribute(@NotNull final Class<?> entityClass,
                                       @NotNull final String name,
                                       @NotNull final Class<?> attributeClass) {
        return entityModel.getTypes()
            .stream()
            .filter(type -> entityClass.equals(type.getClassType()))
            .findFirst()
            .flatMap(type -> getSingleKeyAttribute(type))
            .filter(attr -> attr.getClassType().equals(attributeClass))
            .map(attr -> attr.getName().equals(name))
            .orElse(false);

    }

    @NotNull
    private Collection<Class<?>> getManagedTypes() {
        if (!managedTypes.isPresent()) {
            Set<Type<?>> managedTypes = entityModel.getTypes();
            Set<Class<?>> types = new HashSet<>(managedTypes.size());

            for (Type<?> managedType : managedTypes) {
                Class<?> type = managedType.getClassType();
                if (type != null) {
                    types.add(type);
                }
            }
            this.managedTypes = Optional.of(types);
        }

        return this.managedTypes.get();
    }


    private static Optional<? extends Attribute<?, ?>> getSingleKeyAttribute(Type<?> type) {
        return Optional.ofNullable(type.getSingleKeyAttribute());
    }

}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/utils/RequeryMetamodel.kt
package org.springframework.data.requery.utils

import io.requery.meta.EntityModel

/**
 * RequeryMetamodel
 *
 * @author debop@coupang.com
 */
class RequeryMetamodel(val entityModel: EntityModel) {

    private val managedTypes = arrayListOf<Class<*>>()

    fun isRequeryManaged(entityClass: Class<*>): Boolean {
        return getManagedTypes().contains(entityClass)
    }

    fun isSingleIdAttribute(entityClass: Class<*>, name: String, attributeClass: Class<*>): Boolean {

        return entityModel.types
                   .filter { it.classType == entityClass }
                   .mapNotNull { it.singleKeyAttribute }
                   .filter { it.classType == attributeClass }
                   .map { it.name == name }
                   .firstOrNull()
               ?: false

    }

    private fun getManagedTypes(): Collection<Class<*>> {
        if(managedTypes.isEmpty()) {
            managedTypes.addAll(entityModel.types.mapNotNull { it.classType })
        }
        return managedTypes
    }

}